File: hms-sdk-unity/HuaweiMobileServices/Ads/AdListenerWrapper.cs
namespace HuaweiMobileServices.Ads
{
    using HuaweiMobileServices.Utils;
    using UnityEngine;

    internal class AdListenerWrapper : JavaObjectWrapper
    {

        private class AdListenerInterfaceWrapper : AndroidJavaProxy
        {

            private readonly IAdListener mListener;

            public AdListenerInterfaceWrapper(IAdListener listener) : base("org.m0skit0.android.hms.unity.ads.AdListener")
            {
                mListener = listener;
            }

            public void onAdClosed()
            {
                this.CallOnMainThread(() => { mListener.OnAdClosed(); });
            }

            public void onAdFailed(int paramInt)
            {
                this.CallOnMainThread(() => { mListener.OnAdFailed(paramInt); });
            }

            public void onAdLeave()
            {
                this.CallOnMainThread(() => { mListener.OnAdLeave(); });
            }

            public void onAdOpened()
            {
                this.CallOnMainThread(() => { mListener.OnAdOpened(); });
            }

            public void onAdLoaded()
            {
                this.CallOnMainThread(() => { mListener.OnAdLoaded(); });
            }

            public void onAdClicked()
            {
                this.CallOnMainThread(() => { mListener.OnAdClicked(); });
            }

            public void onAdImpression()
            {
                this.CallOnMainThread(() => { mListener.OnAdImpression(); });
            }
        }

        public AdListenerWrapper(IAdListener listener)
            : base("org.m0skit0.android.hms.unity.ads.AdListenerWrapper", new AdListenerInterfaceWrapper(listener)) { }

    }
}

File: android-wrapper/app/src/main/java/org/m0skit0/android/hms/unity/ads/AdListenerWrapper.java
package org.m0skit0.android.hms.unity.ads;


public class AdListenerWrapper extends com.huawei.hms.ads.AdListener {

    private final AdListener listener;

    public AdListenerWrapper(final AdListener listener) {
        super();
        this.listener = listener;
    }

    @Override
    public void onAdClosed() {
        super.onAdClosed();
        listener.onAdClosed();
    }

    @Override
    public void onAdFailed(int errorCode) {
        super.onAdFailed(errorCode);
        listener.onAdFailed(errorCode);
    }

    @Override
    public void onAdLeave() {
        super.onAdLeave();
        listener.onAdLeave();
    }

    @Override
    public void onAdOpened() {
        super.onAdOpened();
        listener.onAdOpened();
    }

    @Override
    public void onAdLoaded() {
        super.onAdLoaded();
        listener.onAdLoaded();
    }

    @Override
    public void onAdClicked() {
        super.onAdClicked();
        listener.onAdClicked();
    }

    @Override
    public void onAdImpression() {
        super.onAdImpression();
        listener.onAdImpression();
    }
}
